### PR TITLE
Change onBeforeRun API 

### PR DESCRIPTION
**Pull request by jweak**
A hook will be run for all catalog entities and the hook function can decide if it wants to do something with it.

Signed-off-by: Juho Heikka <juho.heikka@gmail.com>

### CODE REVIEW

File: src/extensions/renderer-api/catalog.ts [hunk 1]
```diff
@@ -56,8 +56,8 @@ export class CatalogEntityRegistry {
    * @param onBeforeRun The function that should return a boolean if the onRun of catalog entity should be triggered.
    * @returns A function to remove that hook
    */
-  addOnBeforeRun(entity: CatalogEntity, onBeforeRun: CatalogEntityOnBeforeRun): Disposer {
-    return registry.addOnBeforeRun(entity, onBeforeRun);
+  addOnBeforeRun(onBeforeRun: CatalogEntityOnBeforeRun): Disposer {
```
Comment (jakolehm) [inline src/extensions/renderer-api/catalog.ts:61]: Tsdoc needs update.

Comment (jweak) [reply src/extensions/renderer-api/catalog.ts:61]: Good catch, fixed!

**Review by jakolehm — DISMISSED** [review]

File: src/common/catalog/catalog-run-event.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import type { CatalogEntity } from "../catalog";
+
+export class CatalogRunEvent {
+  private _preventDefaulted: boolean;
```
Comment (Nokel81) [inline src/common/catalog/catalog-run-event.ts]: Might as well just use `#preventDefault`

File: src/extensions/renderer-api/catalog.ts [hunk 3]
```diff
@@ -51,13 +51,15 @@ export class CatalogEntityRegistry {
   }
 
   /**
-   * Add a onBeforeRun hook to a catalog entity. If `onBeforeRun` was previously added then it will not be added again
-   * @param catalogEntityUid The uid of the catalog entity
-   * @param onBeforeRun The function that should return a boolean if the onRun of catalog entity should be triggered.
+   * Add a onBeforeRun hook to a catalog entities. If `onBeforeRun` was previously
+   * added then it will not be added again.
+   * @param onBeforeRun The function will receive the catalog run event as a parameter.
```
Comment (Nokel81) [inline src/extensions/renderer-api/catalog.ts]: How about this wording?
```suggestion
   * @param onBeforeRun The function to be called with a `CatalogRunEvent`
```

File: src/renderer/api/catalog-entity-registry.ts [hunk 4]
```diff
@@ -29,19 +29,18 @@ import { Disposer, iter } from "../utils";
 import { once } from "lodash";
 import logger from "../../common/logger";
 import { catalogEntityRunContext } from "./catalog-entity";
+import { CatalogRunEvent } from "../../common/catalog/catalog-run-event";
 
 export type EntityFilter = (entity: CatalogEntity) => any;
-export type CatalogEntityOnBeforeRun = (entity: CatalogEntity) => boolean | Promise<boolean>;
-
-type CatalogEntityUid = CatalogEntity["metadata"]["uid"];
+export type CatalogEntityOnBeforeRun = ((event: CatalogRunEvent) => void) | ((event: CatalogRunEvent) => Promise<void>);
```
Comment (Nokel81) [inline src/renderer/api/catalog-entity-registry.ts]: This can be simplified
```suggestion
export type CatalogEntityOnBeforeRun = (event: CatalogRunEvent) => void | Promise<void>;
```

File: src/renderer/api/catalog-entity-registry.ts [hunk 5]
```diff
@@ -179,44 +178,35 @@ export class CatalogEntityRegistry {
...
-        if (!await onBeforeRun(entity)) {
-          return false;
-        }
+        const runEvent = new CatalogRunEvent({ target: entity });
+
+        await onBeforeRun(runEvent);
+
+        const shouldContinue = !runEvent.preventDefaulted;
+
+        return shouldContinue;
```
Comment (Nokel81) [inline src/renderer/api/catalog-entity-registry.ts]: This is wrong, we shouldn't stop after the first one.

Also, why do we create a new event for each hook?

**Review by Nokel81 — CHANGES_REQUESTED** [review]

**Review by Nokel81 — DISMISSED** [review]

**Review by Nokel81 — APPROVED** [review]
